### ISSUE TITLE
Difference between BIO_read_filename and BIO_new_file in Openssl 3.0

### ISSUE THREAD

**Issue by casidapulak** [openssl/openssl#21271]
Hello,
When using Openssl 3.0.8, for **d2i_PKCS7_bio** api the following code works:
 ```
    BIO* certBio = BIO_new(BIO_s_file());
    BIO_read_filename(certBio, "FilePath");
    PKCS7* p7b = d2i_PKCS7_bio(certBio, NULL);
 ```
But the following doesnot:
 ```
BIO* bio = BIO_new_file("FIlePath", "r");
PKCS7* p7b = d2i_PKCS7_bio(certBio, NULL);
 ```
I was getting a null pointer. When I checked internally, both seem to be using BIO_CTRL one with File ptr and the other with file name. Can you please let me know what I am missing. Thanks

**Comment by mattcaswell** [comment]
Interesting. I would expect these to be equivalent.

> I was getting a null pointer. 

I assume you mean you were getting a NULL pointer from `d2i_PKCS7_bio()` rather than from `BIO_new_file`?

Is there anything interesting on the OpenSSL error stack? Try inserting this after the failure to dump errors from the error stack to stderr:

````
    ERR_print_errors_fp(stderr);
````

**Comment by casidapulak** [comment]
@mattcaswell 
Yes, the null pointer was coming from d2i_PKCS7_bio().
It is crashing when I use the api  ERR_print_errors_fp(stderr); 
But when I debug the code the error is raised from the below portion of code in a_d2i_fp.c. 
```
                    while (chunk > 0) {
                        i = BIO_read(in, &(b->data[len]), chunk);
                        if (i <= 0) {
                            ERR_raise(ERR_LIB_ASN1, ASN1_R_NOT_ENOUGH_DATA);
                            goto err;
                        }
```
Please let me know if you need any other details that I can provide. Thanks

**Comment by mattcaswell** [comment]
> It is crashing when I use the api ERR_print_errors_fp(stderr);

!!!????

What platform are you running this on?

**Comment by casidapulak** [comment]
windows 32 bit

**Comment by mattcaswell** [comment]
Have you compiled/linked the applink.c file into your application?

https://www.openssl.org/docs/manmaster/man3/OPENSSL_Applink.html

**Comment by casidapulak** [comment]
I think we are not doing that step. That may be the root cause of the issue. I am a bit new to Openssl 3.0 compilation. Can you please point me to any relevant documentation for compiling/linking applink.c into our application.

**Comment by mattcaswell** [comment]
You should find applink.c in the same place as your header files. You just need to compile and link it into your application in the same way you would any other .c file. Documentation is "sparse" (I already gave you the only doc link we have on this).

Although, I'm not entirely sure this is actually the cause of your problem. applink.c problems are normally the result of passing `FILE *` pointers between the application and the library (which is probably why the `ERR_print_errors_fp()` call failed above)...but you don't seem to be doing that in the code snippets that you have provided.

**Comment by casidapulak** [comment]
Ok, after including applink.c in my target application, the ERR_print_errors_fp() is not longer crashing. When I dump stderr, I don't see anything.
Compiling with applink.c still did not solve the original issue though

**Comment by mattcaswell** [comment]
> Another interesting this is when I change the mode of BIO_new_file from "r" to "rb" it seems to be working as expected.

Ah! That actually seems to be the difference with `BIO_read_filename`. When I look at the implementation of the `BIO_C_SET_FILENAME` ctrl on Windows I see it does this:

```` C
# if defined(OPENSSL_SYS_MSDOS) || defined(OPENSSL_SYS_WINDOWS)
        if (!(num & BIO_FP_TEXT))
            OPENSSL_strlcat(p, "b", sizeof(p));
        else
            OPENSSL_strlcat(p, "t", sizeof(p));
````
In other words its going to end up opening it in "rb" mode, which would explain why that version works, but the other one doesn't.

**Comment by casidapulak** [comment]
Thanks, that explains it! But my final question is why it worked in "r" mode in openssl 1.0.2 with the same file. Did some thing change with regards to the way the files are read?

**Comment by mattcaswell** [comment]
> Thanks, that explains it! But my final question is why it worked in "r" mode in openssl 1.0.2 with the same file. Did some thing change with regards to the way the files are read?

Not that I'm aware of....but *lots* of things have changed since 1.0.2, so its entirely possible.

**Comment by nhorman** [comment]
Marking as inactive, to be closed on the end of the 3.4 dev cycle, barring further input
